ckan/views/dataset.py

```diff
@@ -1108,21 +1117,42 @@ def activity(package_type: str, id: str) -> Union[Response, str]:  # noqa
         u'auth_user_obj': g.userobj
     })
     data_dict = {u'id': id}
+    base_limit = int(config.get(u'ckan.activity_list_limit', 30))
+    max_limit = int(config.get(u'ckan.activity_list_limit_max', 100))
+    limit = min(base_limit, max_limit)
+    page = int(page)
+    offset = page * limit
+    activity_types = [activity_type] if activity_type else None
     try:
         pkg_dict = get_action(u'package_show')(context, data_dict)
+        activity_dict = {
+            u'id': pkg_dict[u'id'],
+            u'offset': offset,
+            u'limit': limit,
+            u'activity_types': activity_types
+        }
         pkg = context[u'package']
         package_activity_stream = get_action(
             u'package_activity_list')(
-            context, {u'id': pkg_dict[u'id']})
+                context,
+                activity_dict,
+        )
         dataset_type = pkg_dict[u'type'] or u'dataset'
     except NotFound:
         return base.abort(404, _(u'Dataset not found'))
     except NotAuthorized:
         return base.abort(403, _(u'Unauthorized to read dataset %s') % id)
 
-    # TODO: remove
-    g.pkg_dict = pkg_dict
-    g.pkg = pkg
+    all_activities = activity_model.package_activity_list(
```

Comment:
This looks like a really expensive way of calculating the total number of activities. Maybe we can create a new `package_activity_count` function  that accepts the same params but just returns the count needed, computed in an efficient way

Reply:
I added the `package_activity_count` [here](https://github.com/avdata99/ckan/commit/cc146984c2b4762ea20bcfc03a3e37c244330194) @amercader